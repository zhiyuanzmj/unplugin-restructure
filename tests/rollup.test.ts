import path from 'node:path'
import { rollupBuild, RollupVueJsx, testFixtures } from '@vue-macros/test-utils'
import { describe } from 'vitest'
import Restructure from '../src/rollup'

describe('restructure', async () => {
  const { dirname } = import.meta
  await testFixtures(
    '*.tsx',
    (args, id) => rollupBuild(id, [RollupVueJsx(), Restructure() as any]),
    { cwd: path.resolve(dirname, 'fixtures'), promise: true },
  )
})

describe('restructure unwrapRef', async () => {
  const { dirname } = import.meta
  await testFixtures(
    '*.tsx',
    (args, id) =>
      rollupBuild(id, [
        RollupVueJsx(),
        Restructure({
          unwrapRef: true,
        }) as any,
      ]),
    { cwd: path.resolve(dirname, 'fixtures'), promise: true },
  )
})
