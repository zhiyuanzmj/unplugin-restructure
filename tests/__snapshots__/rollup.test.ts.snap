// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`restructure > restructure.tsx 1`] = `
"// restructure.js
import { createVNode, Fragment, createPropsRestProxy } from 'vue';

function resolveDefaultProps(paths) {
  const result = {};

  for (const path of Object.keys(paths)) {
    const segments = path.split(/[?.[\\]]/).filter(Boolean);
    let current = result;

    for (let i = 0; i < segments.length; i++) {
      const segment = segments[i];
      if (i === segments.length - 1) {
        current[segment] = paths[path];
      } else {
        if (!current[segment]) {
          current[segment] = Number.isNaN(Number(segments[i + 1])) ? {} : [];
        }
        current = current[segment];
      }
    }
  }

  return result
}

function createPropsDefaultProxy(props, defaults) {
  const defaultProps = resolveDefaultProps(defaults);
  const result = {};

  for (const key of [
    ...new Set([...Object.keys(props), ...Object.keys(defaultProps)]),
  ]) {
    Object.defineProperty(result, key, {
      enumerable: true,
      get: () => (props[key] === undefined ? defaultProps[key] : props[key]),
    });
  }

  return result
}

const App = /* @restructure */(__MACROS_props, __MACROS_props1) => {
  function onClick({
    foo
  }) {
    return {
      foo,
      baz: __MACROS_props1.baz.baz
    };
  }
  ;
  return [__MACROS_props[0][0][1], __MACROS_props[1].id.foo[0], __MACROS_props1.baz];
};
const App1 = /* @restructure */(__MACROS_props) => createVNode(Fragment, null, [[__MACROS_props[0].root.foo]]);

/* @restructure */
function App2(__MACROS_props, __MACROS_props1) {
const __MACROS_default_props = createPropsDefaultProxy(__MACROS_props, {'.foo': 'bar'});
const __MACROS_default_props1 = createPropsDefaultProxy(__MACROS_props1, {'[0]': 'foo'});
const rest = createPropsRestProxy(__MACROS_default_props, ['foo', 'baz']);
  return createVNode(Fragment, null, [[__MACROS_default_props.foo, __MACROS_default_props.baz, rest, __MACROS_default_props1[0]]]);
}

/* @restructure */
function App3(__MACROS_props) {
const __MACROS_default_props = createPropsDefaultProxy(__MACROS_props, {'.bar': 1});
const rest = createPropsRestProxy(__MACROS_default_props, ['foo', 'bar']);
  return createVNode(Fragment, null, [[__MACROS_default_props.foo, __MACROS_default_props.bar, rest]]);
}

/* @restructure */
function App4(__MACROS_props) {
const __MACROS_default_props = createPropsDefaultProxy(__MACROS_props, {'[1]': 1});
const rest = createPropsRestProxy(__MACROS_default_props, ['foo', 'bar']);
  return createVNode(Fragment, null, [[__MACROS_default_props[0], __MACROS_default_props[1], rest]]);
}

/* @restructure */
function App5(__MACROS_props, bar) {
  return createVNode(Fragment, null, [[__MACROS_props.foo, bar]]);
}
"
`;

exports[`restructure unwrapRef > restructure.tsx 1`] = `
"// restructure.js
import { createVNode, Fragment, createPropsRestProxy } from 'vue';

function resolveDefaultProps(paths) {
  const result = {};

  for (const path of Object.keys(paths)) {
    const segments = path.split(/[?.[\\]]/).filter(Boolean);
    let current = result;

    for (let i = 0; i < segments.length; i++) {
      const segment = segments[i];
      if (i === segments.length - 1) {
        current[segment] = paths[path];
      } else {
        if (!current[segment]) {
          current[segment] = Number.isNaN(Number(segments[i + 1])) ? {} : [];
        }
        current = current[segment];
      }
    }
  }

  return result
}

function createPropsDefaultProxy(props, defaults) {
  const defaultProps = resolveDefaultProps(defaults);
  const result = {};

  for (const key of [
    ...new Set([...Object.keys(props), ...Object.keys(defaultProps)]),
  ]) {
    Object.defineProperty(result, key, {
      enumerable: true,
      get: () => (props[key] === undefined ? defaultProps[key] : props[key]),
    });
  }

  return result
}

const App = /* @restructure */(__MACROS_props, __MACROS_props1) => {
  function onClick({
    foo
  }) {
    return {
      foo,
      baz: __MACROS_props1.value.baz.baz
    };
  }
  ;
  return [__MACROS_props.value, __MACROS_props.value.id.foo, __MACROS_props1.value.baz];
};
const App1 = /* @restructure */(__MACROS_props) => createVNode(Fragment, null, [[__MACROS_props.value.root.foo]]);

/* @restructure */
function App2(__MACROS_props, __MACROS_props1) {
const __MACROS_default_props = createPropsDefaultProxy(__MACROS_props, {'.value.foo': 'bar'});
const __MACROS_default_props1 = createPropsDefaultProxy(__MACROS_props1, {'.value': 'foo'});
const rest = createPropsRestProxy(__MACROS_default_props.value, ['foo', 'baz']);
  return createVNode(Fragment, null, [[__MACROS_default_props.value.foo, __MACROS_default_props.value.baz, rest, __MACROS_default_props1.value]]);
}

/* @restructure */
function App3(__MACROS_props) {
const __MACROS_default_props = createPropsDefaultProxy(__MACROS_props, {'.value.bar': 1});
const rest = createPropsRestProxy(__MACROS_default_props.value, ['foo', 'bar']);
  return createVNode(Fragment, null, [[__MACROS_default_props.value.foo, __MACROS_default_props.value.bar, rest]]);
}

/* @restructure */
function App4(__MACROS_props) {
const __MACROS_default_props = createPropsDefaultProxy(__MACROS_props, {'.value': 1});
const rest = createPropsRestProxy(__MACROS_default_props.value, ['foo', 'bar']);
  return createVNode(Fragment, null, [[__MACROS_default_props.value, __MACROS_default_props.value, rest]]);
}

/* @restructure */
function App5(__MACROS_props, __MACROS_props1) {
  return createVNode(Fragment, null, [[__MACROS_props.value.foo, __MACROS_props1.value]]);
}
"
`;
